File: aoc2025-go/parse/parse.go
package parse

import (
	"fmt"
	"os"
	"strings"
)

func path(day int, isTest bool) string {
	if isTest {
		return fmt.Sprintf("../data/day%02d_test.txt", day)
	}
	return fmt.Sprintf("../data/day%02d.txt", day)
}

func String(day int, isTest bool) string {
	p := path(day, isTest)
	data, err := os.ReadFile(p)
	if err != nil {
		fmt.Fprintf(os.Stderr, "missing input: %s\n", p)
		os.Exit(1)
	}
	return strings.TrimSpace(string(data))
}

func Lines(day int, isTest bool) []string {
	return strings.Split(String(day, isTest), "\n")
}

func Sections(day int, isTest bool) [][]string {
	raw := String(day, isTest)
	parts := strings.Split(raw, "\n\n")
	sections := make([][]string, len(parts))
	for i, part := range parts {
		sections[i] = strings.Split(strings.TrimSpace(part), "\n")
	}
	return sections
}

func PrintString(s string) {
	fmt.Println(s)
}

func PrintLines(lines []string) {
	for _, line := range lines {
		fmt.Printf("%s\n", line)
	}
}

File: aoc2025-zig/src/parse.zig
const std = @import("std");
const common = @import("common.zig");

pub fn string(io: std.Io, allocator: std.mem.Allocator, day: u8, is_test: bool) ![]const u8 {
    var path_buf: [48]u8 = undefined;
    const suffix: []const u8 = if (is_test) "_test" else "";
    const path = try std.fmt.bufPrint(&path_buf, "../data/day{d:0>2}{s}.txt", .{ day, suffix });
    const raw = std.Io.Dir.cwd().readFileAlloc(io, path, allocator, .unlimited) catch |err| {
        std.debug.print("missing input: {s}\n", .{path});
        return err;
    };
    return std.mem.trimEnd(u8, raw, "\n\r \t");
}

pub fn lines(io: std.Io, allocator: std.mem.Allocator, day: u8, is_test: bool) ![][]const u8 {
    const raw = try string(io, allocator, day, is_test);
    return common.getStringArray(allocator, raw);
}

pub fn sections(io: std.Io, allocator: std.mem.Allocator, day: u8, is_test: bool) ![][][]const u8 {
    const raw = try string(io, allocator, day, is_test);
    var result: std.ArrayList([][]const u8) = .empty;
    var block_it = std.mem.splitSequence(u8, raw, "\n\n");
    while (block_it.next()) |block| {
        const trimmed = std.mem.trim(u8, block, "\n\r \t");
        const block_lines = try common.getStringArray(allocator, trimmed);
        try result.append(allocator, block_lines);
    }
    return result.toOwnedSlice(allocator);
}

pub fn printString(s: []const u8) void {
    std.debug.print("{s}\n", .{s});
}

pub fn printLines(ls: [][]const u8) void {
    for (ls, 0..) |line, i| {
        std.debug.print("{d}: {s}\n", .{ i, line });
    }
}
